src/main/java/io/github/vcuswimlab/stackintheflow/model/L2H/L2HPredictor.java

```diff
@@ -0,0 +1,182 @@
+	private static final int sampleLag = 25;
+	private static final int alpha = 10;
+	private static final int beta = 1000;
+	private static final int a0 = 90;
+	private static final int b0 = 10;
+
+	private static int[] toIntArray(List<Integer> list){
+	  int[] ret = new int[list.size()];
+	  for(int i = 0;i < ret.length;i++)
+		ret[i] = list.get(i);
+	  return ret;
+	}
+
+	private static class TextDataLoader extends TextDataset {		
+		public TextDataLoader(String dataset, String fileName) throws FileNotFoundException, IOException, Exception {
+			super(dataset);
+			File testFile = new File(fileName);
+			inputTextData(testFile);
+		}
+		
+		public int[][] getWords() {
+			return words;
+		}
+	}
+
+	private static double[][] computeInitPredictions(int[][] newWords, int numUniqTags) {
+        double[][] initPredictions = new double[newWords.length][];
+        double p = 1./ (double) numUniqTags;
+        for (int i=0; i<newWords.length; i++) {
+            initPredictions[i] = new double[numUniqTags];
+            for (int j=0; j<numUniqTags; j++) {
+                initPredictions[i][j] = p;
+            }
+        }
+        return initPredictions;
+    }
+
+	private static class TagProbability implements Comparable<TagProbability> {
+		public String tagName;
+		public double probability;
+
+		public TagProbability(double probability, String tagName) {
+			this.probability = probability;
+			this.tagName = tagName;
+		}
+
+		@Override
+		public int compareTo(TagProbability other) {
+			return Double.compare(other.probability, probability);
+		}
+	}
+
+	private Map<String, Integer> nameToIdMap;
+	private CustomL2H testSampler;
+	private LabelTextDataset data;
+
+	public L2HPredictor() throws IOException {
+		nameToIdMap = new HashMap<>();
+		BufferedReader br = new BufferedReader(new FileReader(formatFolder+formatFile+".wvoc"));
+		int maxId = -1;
+		try {
+			String line;
+			int intId = 0;
+			while ((line = br.readLine()) != null) {
+				nameToIdMap.put(line, intId);
+				intId++;
+			}
+		} finally {
+			br.close();
+		}
+
+		data = new LabelTextDataset(datasetName);
+        data.setFormatFilename(formatFile);
+        data.loadFormattedData(formatFolder);
+        data.filterLabelsByFrequency(minLabelFreq);
+		data.prepareTopicCoherence(numTopWords);
+
+		testSampler = new CustomL2H();
+		testSampler.setVerbose(true);
+        testSampler.setDebug(false);
+        testSampler.setLog(false);
+        testSampler.setReport(false);
+        testSampler.configure("./output", data.getWordVocab().size(), alpha, beta, a0, b0,
+				new MSTBuilder(data.getLabels(), data.getLabelVocab()), false, false, AbstractSampler.InitialState.PRESET, PathAssumption.MAXIMAL, false, burnIn, modelIterations, sampleLag, 1);
+        testSampler.setTestConfigurations(burnIn, modelIterations, sampleLag);
+	}
+
+	public void computeL2HPredictions(String fileToPredict, String outputFileName, String tempFileName) throws Exception {
```

Comment:
I feel like instead of returning void these methods (computeL2HPredictions and computeMostLikelyTags) should be returning collections of some kind.

Reply:
I agree. computeMostLikelyTags is now returning a collection. computeL2HPredictions is still void for now because returning a collection there requires solving the 'writing predictions to temp file and reading back issue' that we discussed earlier.